homeassistant/components/mqtt/client.py

```diff
@@ -515,8 +517,16 @@ async def async_subscribe(
 
         # Only subscribe if currently connected.
         if self.connected:
-            self._last_subscribe = time.time()
-            await self._async_perform_subscriptions(((topic, qos),))
+            if already_active_subscription:
+                _LOGGER.debug("Skipped duplicate subscribe for %s", topic)
+                if topic in self._last_message:
+                    self.hass.async_run_hass_job(
+                        subscription.job, self._last_message.get(topic)
+                    )
+                    self._mqtt_data.state_write_requests.process_write_state_requests()
```

Comment:
This is not correct, it means we treat non-retained MQTT messages as if they are retained.

Reply:
> It would be great if we could avoid subscribing to the same topic multiple times

It's not just great.  It's necessary, this resolves two major issues.

First, is the absolute flood of messages at start up caused by multiple subscribes. In my instance, that's more than 8000 subscribe requests to a single topic (the zigbee2mqtt availability topic) - and that's on TOP of the multiple subscriptions for each entity.  This causes the startup of HASS to be more than 10 minutes, and many of these requests time out or don't get processed (or worse, re-processed for every entity for each subscription).
Second, related, is that the subscriptions are going out so much we are (I assume) getting throttled?  Not sure, but many of the subscription requests don't actually return anything.  So we don't actually get subscribed for many entities, meaning those entities aren't listening on the topics they need to (despite the MQTT being subscribed to it), and therefore don't get updates.

Without this patch, HA MQTT is completely broken.  And I don't mean that as hyperbole.  I mean, in my instance, with only 100 or so Zigbee devices feeding into MQTT (Which ends up being more than 8000 entities, for some reason - though about half of them I think are inactive/should be removed because they were renamed - so let's say 4000) - the MAJORITY of the entities NEVER get updates due to this patch because of the subscribe calls being too many and never returning.  And startup takes 10 minutes waiting for MQTT to issue all these duplicate subscribes.

> When sending a PUBLISH Packet to a Client the Server MUST set the RETAIN flag to 1 if a message is sent as a result of a new
> subscription being made by a Client [MQTT-3.3.1-8]. It MUST set the RETAIN flag to 0 when a PUBLISH Packet is sent to a Client because it matches an established subscription **regardless of how the flag was set in the message it received** [MQTT-3.3.1-9].

The part of the document you pasted explicitly says the server is dropping and ignoring the value of the PUBLISH flag it received from the client - and setting it based on whether the message is sent on an existing or new subscription.  Therefore we don't NEED to worry about what the original publisher of the message set the RETAIN flag.  I added emphasis to show this.

By my reading of this, this is the flow.

Client A subscribes to foo - get message with the RETAIN flag set to 1
Client B subscribes to foo, client A and client B both get a message, but for client A the RETAIN flag is set to 0, and for client B it's set to 1.
Client C subscribes to foo, client A, B and C get a message, but for clients A and B the RETAIN flag is set to 0, and for client C it's set to 1.
This is all true regardless of what the publisher of the original message set the RETAIN flag.

In this case, I think we are fine either way.  I would say that we should consider HA MQTT the client, not every individual entity HA MQTT provides to HASS.  Meaning that, with this, HA MQTT subscribes once, and if something else (something other than HASS) subscribed to those topics, we would see messages with RETAIN=0 when they do.

This makes the most sense - since @jbouwh already said that, when we send a new subscribe message, we are not subscribing a second time - MQTT will only allow one subscription per client per topic - and client in this case is indicated as a single socket connection between HA MQTT and the MQTT server.

Since we don't create/configure all entities first, THEN subscribe to the topics for each entity, but instead subscribe when we first see the entity, we are just informing each new "subscriber" of the current data for that topic.  From a MQTT perspective, each entity is **not** a new client, and we should not be re-subscribing - something that MQTT will not allow any (one subscription per client) - we are just needlessly flooding the MQTT server with subscribe requests, and ourselves with duplicate responses.

It's not like we'd miss updates - once we're subscribed, we will be getting the updated information reardless.

Reply:
So I guess the exception to this is:

- Client sends message A with RETAIN 1
- MQTT sends message A to all other clients (with retain set above)
- Client sends message B with RETAIN 0
- MQTT sends message B to all other clients (with retain set above)
- New entity subscribes to topic, and should receive message A (not B) because THAT was what was the retained message in MQTT.  However, with this patch, they will receive message B.


There is no way we can get around this, that's true, as we don't know what the original RETAIN value was - but I would argue that this is not a frequent case, and more importantly, if we DON'T implement this change, the HA MQTT server is unusable for people who have a high volume of topics or high volume of entities subscribed to the same topic (which is the case for Zigbee2MQTT users, as EVERY entity created subscribes to the same status topic).

Even the CURRENT behavior we use is wrong.

Right now, this is the current behavior - without this PR:

- External client sends message A with RETAIN=1
- We subscribe a bunch of times, for new entities.  Each time we subscribe, EVERY entity we have subscribed so far gets message A. So if we subscribe 5 entities, the first entity gets message 1 5 times, the next 4 times, the next 3 times, etc.
- External client sends message B with RETAIN=0
- All 5 of our subscribed entities get message B
- We subscribe a sixth entity, and then all 6 entities now get blasted with message A again.

We should never have sent message A to the first 5 entities that were already subsribed with the subscription of entity 6, they already got message B (ephemeral as it was), and had already processed message A.  But now they will all be re-processing the retained message, despite the fact that, if they were all truly separate MQTT clients (as IT defines client), they would never have been re-issued a message A publish.


There are two potential ways around this issue:
1. We delay ALL subscriptions until discovery is complete.  We create the internal subscriptions, so that when we DO do the MQTT subscription, they will all get the messages, but we don't actually subscribe until we have completed processing of the initial blast of discovery messages.
2. Every N seconds, allow a new subscribe message to go through on a new subscription, which would trigger sending message A to everyone who "fake subscribed" and got message B (but not those who already got 'A'.

The drawback to option 1 is that it's a much bigger solution, and I'd rather fix the completely broken state of HA-MQTT for high-entity clients than wait on that.  And possibly have doing that as a future task.  Plus it means we won't get any data for any entity until discovery and entity creation is over (not necessarily a bad thing).  That said though, we have no real way of knowing when discovery is over.
The drawback for option 2 is it's  hack, and not reliable.

Reply:
FYI - here is what the code to not spam previously subscribed topics looks like
https://github.com/corporategoth/core/pull/2

But I don't think this is the correct way to go - because although it DOES fix the extremely slow startup problem (not re-processing the initial messages for all new subscreiptions) - it doesn't fix the spamming MQTT repeatedly for the same topic problem.  And this problem is what is causing a large percentage of my existing entities to fail - the subscription request for them is dropped on the floor (either by MQTT or HA or something).

Reply:
Subscribing to the same topic is effectively O(n²) performant. That's not great, as it doesn't scale well.

As this PR stands now, it's a significant improvement for one set of users. However, there is a risk of collateral damage, and we are trading one problem for another, which may create a fire for the maintainers of this integration to put out, making everyone reluctant to want to merge this in the current state.

You have put a lot of thought into this. Please find a solution to address Erik's feedback above so we can merge this.

Reply:
@bdraco I mentioned 3 different "solutions" - along with the drawbacks for each, addressing Erik's feedback.  And have not gotten feedback on any one of them.

> "Please solve P=NP before we will think about patching the leak in our boat making it unseaworthy."
> "Here are some potential problems with drawbacks, what are your thoughts?"
> "Solve P=NP before we fix the boat!"
> "That does not give me any information to go on."
> "Just come up with a solution!  I'm not going to help you!  Use magic to get the perfect solution!  We won't accept anything less.  But hurry, because we really WOULD like to fix this leaky boat."
> "..."

I submitted this MR to solve an emergent (and debilitating to the point of the MQTT being broken at any reasonable scale) problem, mainly because @jbouwh kept closing my bug reports as "not going to fix" or "probably zigbee2mqtts problem".  I wasn't planning on re-designing how the mqtt module works - which I would have to do to fix Erik's issue.  And I am certainly not going to do that without feedback from the people who have been working on it more than the 48 hours I have.

It's not perfect.  It doesn't handle every case.  It may need to change with a later re-design to implement one of the solutions I mentioned above.  But that's what iterative development is for.  But it does stop MQTT from being completely useless at any kind of scale (at least when used with zigbee2mqtt, which uses the same availability topic for every entity - which is a valid approach).  And hell, having zigbee devices entities, like I do right now, isn't even that outrageous - and yet it's enough to break this integration completely.

But let's be clear - ANY of the solutions will probably take another week to solve (half of which is both figuring out the rest of how the MQTT module works, and deciphering the unit tests).  Which you're asking me to do, someone not a maintainer of this module, or any part of home-assistant.  I'm more inclined to just keep patching client.py on my local instance.

Reply:
Why do you think I mentioned P=NP?  There *IS* no perfect solution.

The best solution I think so far is likely delaying any subscriptions until discovery has discovered all topics to be subscribed to.  But that, as you say, has some drawbacks too.

But for now, I'm going to leave it.  I have a solution that solves the most critical problem, even if it doesn't solve everything.  I am a strong believer in "Perfect is the enemy of good."  So I guess MQTT integration will remain broken until I have enough time to dedicate to finding the "perfect" solution.  Or someone who actually knows how the code works looks at it.  Or gives me actual feedback that's not "Hmmm.. interesting .. but everything you mentioned is not perfect.  Keep banging that against the wall - you'll break through someday!"